solutions/gold/ioi-11-crocodile.mdx

```diff
@@ -0,0 +1,94 @@
+---
+id: ioi-11-crocodile
+source: IOI
+title: 2011 - Crocodile
+author: Justin Ji
+---
+
+[Official Analysis](https://dmoj.ca/problem/ioi11p4/editorial)
+
+<Spoiler title="Hint 1">
+
+Instead of trying to traverse from our starting node to an endpoint, consider
+starting from our endpoints and try to reach our starting node.
+
+</Spoiler>
+
+<Spoiler title="Hint 2">
+
+Obviously, the crocodile will try to prevent us from taking the best path from
+a given endpoint to any other node. How can we use this knowledge to decide the
+shortest time it would take to reach a node from any endpoint?
+
+</Spoiler>
+
+<Spoiler title="Solution">
+
+## Explanation
+
+Say we want to reach a given node from an endpoint in our graph. The crocodile
+will always block the best route to an endpoint in our graph, which means we
+have to take the second best route to an endpoint.
+
+Handling decision making for what the "best choice" at a given vertex is
+difficult if we are to implement a shortest paths algorithm directly to our
+graph. Thus, we instead choose to model our problem in a slightly different way:
+instead of considering the best distance from our starting node to every endpoint,
+we consider the best distance from any endpoint to our starting node.
+
+This makes handling decision making significantly easier because it means that
+we can only use the second best distance to any given node. Our algorithm then
+becomes your typical multi-source Dijkstra, except we push to the priority queue
+every time the second best distance becomes better.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(M\log{M})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int travel_plan(int n, int m, int r[][2], int l[], int k, int p[]) {
+	vector<vector<array<int, 2>>> adj(n);
+	for (int i = 0; i < m; i++) {
+		adj[r[i][0]].push_back({r[i][1], l[i]});
+		adj[r[i][1]].push_back({r[i][0], l[i]});
+	}
+
+	// we use 1e9 instead of INT_MAX to prevent overflow
+	const array<int, 2> DEF = {(int)1e9, (int)1e9};
+	vector<array<int, 2>> dist(n, DEF);
+	priority_queue<array<int, 2>> pq;
+	for (int i = 0; i < k; i++) {
+		pq.push({0, p[i]});
+		dist[p[i]] = {0, 0};
+	}
+
+	while (!pq.empty()) {
+		auto [time, at] = pq.top();
+		pq.pop();
+		time *= -1;  // undoing the negative number trick
+		if (dist[at][1] < time) { continue; }
+		for (const auto [nxt, weight] : adj[at]) {
```

Comment:
you can do unpacking with arrays????

Reply:
yeah u can do structured bindings on `std::array` lol